wire-schema/src/main/java/com/squareup/wire/schema/IdentifierSet.java

```diff
@@ -15,73 +15,100 @@
  */
 package com.squareup.wire.schema;
 
-import java.util.NavigableSet;
-import java.util.TreeSet;
+import com.google.common.collect.ImmutableSet;
 
 /**
- * A heterogeneous set of type names and type members. If a member is added to the set, its type is
- * implicitly also added. A type that is added without a specific member implicitly contains all
- * of that type's members.
+ * A heterogeneous set of type names and type members. If a member is included in the set, its type
+ * is implicitly also included. A type that is included without a specific member implicitly
+ * includes all of that type's members.
  *
- * <p>For example, an identifiers set populated with {@code Movie} and {@code Actor#name} contains
- * all members of {@code Movie} (such as {@code Movie#name} and {@code Movie#release_date}). It
- * contains the type {@code Actor} and one member {@code Actor#name}, but not {@code
- * Actor#birth_date} or {@code Actor#oscar_count}.
+ * <p>Identifiers in this set may be in the following forms:
+ * <ul>
+ *   <li>Fully qualified type names, like {@code squareup.protos.person.Person}.
+ *   <li>Fully qualified type names, followed by a '#', followed by a member name, like
+ *       {@code squareup.protos.person.Person#address}.
+ *   <li>Fully qualified service names, like {@code com.squareup.services.ExampleService}.
+ *   <li>Fully qualified service names, followed by a '#', followed by an RPC name, like
+ *       {@code com.squareup.services.ExampleService#SendSomeData}.
+ * </ul>
+ *
+ * <p>An identifiers set populated with {@code Movie} and {@code Actor#name} contains all members of
+ * {@code Movie} (such as {@code Movie#name} and {@code Movie#release_date}). It contains the type
+ * {@code Actor} and one member {@code Actor#name}, but not {@code Actor#birth_date} or {@code
+ * Actor#oscar_count}.
+ *
+ * <p>This set is initialized with <i>included identifiers</i> and <i>excluded identifiers</i>, with
+ * excludes taking precedence over includes. That is, if a type {@code Movie} is in both the
+ * includes and the excludes, it is not contained in the set. Any attempt to mark an excluded
+ * identifier will return false, preventing the identifier's transitive dependencies from also being
+ * marked.
+ *
+ * <p>If the includes set is empty, that implies that all elements should be included. Use this to
+ * exclude unwanted types and members without also including everything else.
  */
 public final class IdentifierSet {
```

Comment:
public API! And a good place to document the roots syntax.

(Will probably rename roots to includes and excludes, in follow-up.)